File: source/uwp/Renderer/lib/DateTimeParser.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
#include "pch.h"
#include "DateTimeParser.h"
#include "Util.h"
#include <iomanip>
#include <sstream>

namespace AdaptiveNamespace
{
    DateTimeParser::DateTimeParser(const std::string& language) { m_languageString = language; }

    std::string DateTimeParser::GenerateString(const DateTimePreparser& text)
    {
        std::ostringstream parsedostr;

        if (text.HasDateTokens())
        {
            std::locale language;
            try
            {
                language = std::locale(m_languageString.c_str());
            }
            catch (...)
            {
                language = std::locale("");
            }
            parsedostr.imbue(language);
        }

        for (const auto& textSection : text.GetTextTokens())
        {
            std::tm result = {};

            result.tm_mday = textSection->GetDay();
            result.tm_mon = textSection->GetMonth();
            result.tm_year = textSection->GetYear() >= 1900 ? textSection->GetYear() - 1900 : 0;

            // using the put_time function the 3 formats are locale dependent
            switch (textSection->GetFormat())
            {
            case DateTimePreparsedTokenFormat::DateCompact:
                parsedostr << std::put_time(&result, "%Ex");
                break;
            case DateTimePreparsedTokenFormat::DateShort:
                mktime(&result);
                parsedostr << std::put_time(&result, "%a, %b %e, %Y");
                break;
            case DateTimePreparsedTokenFormat::DateLong:
                mktime(&result);
                parsedostr << std::put_time(&result, "%A, %B %e, %Y");
                break;
            case DateTimePreparsedTokenFormat::RegularString:
            default:
                parsedostr << textSection->GetText();
                break;
            }
        }

        return parsedostr.str();
    }

}

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/DateTimeParser.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
package io.adaptivecards.renderer.readonly;

import io.adaptivecards.objectmodel.DateTimePreparsedToken;
import io.adaptivecards.objectmodel.DateTimePreparsedTokenFormat;
import io.adaptivecards.objectmodel.DateTimePreparsedTokenVector;
import io.adaptivecards.objectmodel.DateTimePreparser;

import java.text.ParsePosition;
import java.util.Locale;
import java.util.Date;
import java.text.SimpleDateFormat;

public class DateTimeParser {

    public DateTimeParser(String language){
        m_language = new Locale(language);
    }

    public String GenerateString(DateTimePreparser text){
        String parsedString = new String();
        DateTimePreparsedTokenVector textSections = text.GetTextTokens();
        long textSectionLength = textSections.size();

        for(int i = 0; i < textSectionLength; ++i){
            DateTimePreparsedToken ts = textSections.get(i);
            // Java complains of the enum not being a constant type, so have to do this with if elses
            Date dateToParse = FormatDate(ts.GetYear(), ts.GetMonth(), ts.GetDay());
            if (ts.GetFormat() == DateTimePreparsedTokenFormat.DateCompact) {
                SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy");
                parsedString += dateFormat.format(dateToParse);
            } else {
                if (ts.GetFormat() == DateTimePreparsedTokenFormat.DateShort) {
                    SimpleDateFormat localizedFormat = new SimpleDateFormat("EEE, MMM dd, yyyy", m_language);
                    parsedString += localizedFormat.format(dateToParse);
                } else {
                    if (ts.GetFormat() == DateTimePreparsedTokenFormat.DateLong) {
                        SimpleDateFormat localizedFormat = new SimpleDateFormat("EEEE, MMMM dd, yyyy", m_language);
                        parsedString += localizedFormat.format(dateToParse);
                    } else {
                        parsedString += ts.GetText();
                    }
                }
            }
        }

        return parsedString;
    }

    private Date FormatDate(int year, int month, int day){
        SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy");
        try{
            String rawDate = (month + 1) + "/" + day + "/" + year;
            return dateFormat.parse(rawDate, new ParsePosition(0));
        } catch(Exception e){
            return new Date();
        }
    }

    private Locale m_language;

}
